### ISSUE TITLE
Performance of the BPR and WRMF in version 2.0

### ISSUE THREAD

**Issue by ChenSuL** [guoguibing/librec#251]
I find the performance of BPR and WRMF in version 2.0 is a little worse than that in previous version 1.3. 

In the version 1.3, I can reproduce the results in the document “LibRec Examples on Real Data Sets”. The performance of BPR and WRMF in ml-100k dataset is shown in the following table：
![default](https://user-images.githubusercontent.com/18046763/38466934-8be4363a-3b63-11e8-8c06-bc8d9626f0c5.JPG)

In version 2.0, I use the same parameter configuration. The Pre@5 of BPR is **0.364** and the Pre@5 of WRMF is **0.406**.

**Comment by SunYatong** [comment]
There are several reasons that could give rise to these differences:
- The basic structure of 2.0 changed a lot.
- The implementation of the algorithm is a little different.
- The split data could be a little different.

**Comment by ChenSuL** [comment]
I test the rating prediction algorithms (GlobalAvg, UserAvg, ItemAvg, UserKNN, ItemKNN and PMF) of version 2.0 on datasets ml-100k and ml-1m. Under the same parameter configuration，I can get the same results as librec 1.3.  
I test the item recommendation algorithms (MostPop, ItemKNN, UserKNN) on dataset ml-100k. I can also obtain the sames results as librec 1.3.
So, I think the differences of basic structure or data spliting procedure may not be the main reason.

As for the implementation of BPR, I haven't found any differences that can explain the phenomenon. 

In addition, I think the difference between the results is relatively large. For BPR, the difference in PRE@5 is 4%. For WRMF, the difference in PRE@5 is 5%. So I think there should be some other reasons.

**Comment by SunYatong** [comment]
I printed the performance at every iteration, and it reaches the results of 1.3 after 110 iterations.

Results of Top10:
![image](https://user-images.githubusercontent.com/17143022/38780074-da278f22-4103-11e8-9e1d-7f8c7662b63d.png)

Results of Top5:
![image](https://user-images.githubusercontent.com/17143022/38780148-e049e160-4104-11e8-9247-5e539eb1e888.png)

Here is my configuration:

```
rec.recommender.class=bpr
rec.iterator.learnrate=0.05
rec.iterator.learnrate.maximum=0.05
rec.iterator.maximum=300
rec.user.regularization=0.01
rec.item.regularization=0.01
rec.factor.number=10
rec.learnRate.bolddriver=false
rec.learnRate.decay=1.0
rec.recommender.isranking=true
rec.recommender.ranking.topn=10  # or 5
```

**Comment by ChenSuL** [comment]
Yes，you are right. This may be because the algorithm does not converge
But when iteration number is set 110, the bpr in librec 1.3 can obtain higher precision. You can test it.

From your results, I find the metric Recall@5 is very similar to version 1.3.  
So, I think it's may just because the metric PRE@5 is not accurate enought. 

Thank you

**Comment by ChenSuL** [comment]
In order to improve the execution speed of the algorithm, especially in dealing with the big data set, I made some modifications to the BPR implementation, you can test it.

```
/**
 * Copyright (C) 2016 LibRec
 * <p>
 * This file is part of LibRec.
 * LibRec is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * <p>
 * LibRec is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * <p>
 * You should have received a copy of the GNU General Public License
 * along with LibRec. If not, see <http://www.gnu.org/licenses/>.
 */
package net.librec.recommender.cf.ranking;

import net.librec.annotation.ModelData;
import net.librec.common.LibrecException;
import net.librec.math.algorithm.Maths;
import net.librec.math.algorithm.Randoms;
import net.librec.recommender.MatrixFactorizationRecommender;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


/**
 * Rendle et al., <strong>BPR: Bayesian Personalized Ranking from Implicit Feedback</strong>, UAI 2009.
 *
 * @author GuoGuibing and Keqiang Wang
 */
@ModelData({"isRanking", "bpr", "userFactors", "itemFactors"})
public class BPRRecommender extends MatrixFactorizationRecommender {

    @Override
    protected void setup() throws LibrecException {
        super.setup();
    }

    @Override
    protected void trainModel() throws LibrecException {
        
        Map<Integer, List<Integer>> UserItemsSet = new HashMap<Integer, List<Integer>>();
		for(int userIdx=0; userIdx<numUsers; userIdx++)
		{
			UserItemsSet.put(userIdx, trainMatrix.getColumns(userIdx));
		}
        

        for (int iter = 1; iter <= numIterations; iter++) {

            loss = 0.0d;
            for (int sampleCount = 0, smax = numUsers * 100; sampleCount < smax; sampleCount++) {

                // randomly draw (userIdx, posItemIdx, negItemIdx)
                int userIdx, posItemIdx, negItemIdx;
                
                while (true) {
                    userIdx = Randoms.uniform(numUsers);
                    List<Integer> itemList = UserItemsSet.get(userIdx);
                    if (itemList.size() == 0 || itemList.size() == numItems)
                        continue;

                    posItemIdx = itemList.get(Randoms.uniform(itemList.size()));
                    do {
                        negItemIdx = Randoms.uniform(numItems);
                    } while (trainMatrix.get(userIdx, negItemIdx)!=0);

                    break;
                }
         	
                // update parameters
                double posPredictRating = predict(userIdx, posItemIdx);
                double negPredictRating = predict(userIdx, negItemIdx);
                double diffValue = posPredictRating - negPredictRating;

                double lossValue = -Math.log(Maths.logistic(diffValue));
                loss += lossValue;

                double deriValue = Maths.logistic(-diffValue);

                for (int factorIdx = 0; factorIdx < numFactors; factorIdx++) {
                    double userFactorValue = userFactors.get(userIdx, factorIdx);
                    double posItemFactorValue = itemFactors.get(posItemIdx, factorIdx);
                    double negItemFactorValue = itemFactors.get(negItemIdx, factorIdx);

                    userFactors.add(userIdx, factorIdx, learnRate * (deriValue * (posItemFactorValue - negItemFactorValue) - regUser * userFactorValue));
                    itemFactors.add(posItemIdx, factorIdx, learnRate * (deriValue * userFactorValue - regItem * posItemFactorValue));
                    itemFactors.add(negItemIdx, factorIdx, learnRate * (deriValue * (-userFactorValue) - regItem * negItemFactorValue));

                    loss += regUser * userFactorValue * userFactorValue + regItem * posItemFactorValue * posItemFactorValue + regItem * negItemFactorValue * negItemFactorValue;
                }
            }
            if (isConverged(iter) && earlyStop) {
                break;
            }
            updateLRate(iter);
        }
    }

    
}
```

**Comment by SunYatong** [comment]
Thanks for your optimization, our new version has improved the code already. Thank you all the same.

**Comment by ChenSuL** [comment]
I find the cause of the problem is the setting of "bolddriver". There is a mistake in the file "bpr-test.properties".

The correct configuration should be：
**rec.learnrate.bolddriver=true
rec.learnrate.decay=1.0**
Not to be:
**rec.learnRate.bolddriver=true
rec.learnRate.decay=1.0**

Now, I can reproduce the results in the document.

Thank you!

**Comment by ChenSuL** [comment]
Could you tell me how to printe the performance at every iteration ?

**Comment by SunYatong** [comment]
Just call "recommend" function and the evaluators in each iteration during training.

**Comment by dinry** [comment]
how to run BPRrecommend.java and get eval?
